packages/auth/amplify_auth_cognito_dart/lib/src/state/machines/fetch_auth_session_state_machine.dart

```diff
@@ -279,27 +331,109 @@ class FetchAuthSessionStateMachine extends FetchAuthSessionStateMachineBase {
+        userSubResult = AuthResult.error(authException);
+      }
+    } else {
+      if (userPoolTokens != null) {
+        userPoolTokensResult = AuthResult.success(userPoolTokens);
+        userSubResult = AuthResult.success(userPoolTokens.userId);
+      } else {
+        userPoolTokensResult = const AuthResult.error(
+          SignedOutException.noUserSignedIn(),
+        );
+        userSubResult = const AuthResult.error(
+          SignedOutException.noUserSignedIn(),
+        );
+      }
     }
-    if (event.refreshAwsCredentials) {
-      final idToken = userPoolTokens?.idToken.raw;
-      final awsCredentialsResult = await _retrieveAwsCredentials(
-        existingIdentityId: identityId,
-        federatedIdentity:
-            idToken == null ? null : _FederatedIdentity.cognito(idToken),
+
+    var identityId = result.data.identityId;
+    var awsCredentials = result.data.awsCredentials;
+
+    final hasIdentityPool = _identityPoolConfig != null;
+
+    if (!hasIdentityPool) {
+      credentialsResult = const AuthResult<AWSCredentials>.error(
+        InvalidAccountTypeException.noIdentityPool(),
+      );
+      identityIdResult = const AuthResult<String>.error(
+        InvalidAccountTypeException.noIdentityPool(),
       );
-      identityId = awsCredentialsResult.identityId;
-      awsCredentials = awsCredentialsResult.awsCredentials;
+    } else if (event.refreshAwsCredentials) {
+      final idToken = userPoolTokens?.idToken.raw;
+      try {
+        final awsCredentialsResult = await _retrieveAwsCredentials(
+          existingIdentityId: identityId,
+          federatedIdentity:
+              idToken == null ? null : _FederatedIdentity.cognito(idToken),
+        );
+        identityId = awsCredentialsResult.identityId;
+        awsCredentials = awsCredentialsResult.awsCredentials;
+        credentialsResult = AuthResult.success(awsCredentials);
+        identityIdResult = AuthResult.success(identityId);
+      } on AuthException catch (e) {
+        credentialsResult = AuthResult.error(e);
+        identityIdResult = AuthResult.error(e);
+      } on AWSHttpException catch (e) {
+        final exception = NetworkException(
+          'Refreshing credentials failed due to a network error.',
+          underlyingException: e,
+        );
+        credentialsResult = AuthResult.error(exception);
+        identityIdResult = AuthResult.error(exception);
+      } on Object catch (e) {
+        final authException = UnknownException(
+          'An unknown exception occurred.',
+          underlyingException: e,
+        );
+        credentialsResult = AuthResult.error(authException);
+        identityIdResult = AuthResult.error(authException);
+      }
+    } else {
+      credentialsResult = awsCredentials != null
+          ? AuthResult.success(awsCredentials)
+          : const AuthResult.error(
+              UnknownException(
+                'awsCredentials is null, but refreshAwsCredentials was false.',
```

Comment:
Applies to all instances of `UnknownException` in this file: Let's force unwrap unless we can think of a case where this would be true and prove it through a test.